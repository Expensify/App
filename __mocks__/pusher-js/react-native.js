import {PusherMock} from 'pusher-js-mock';

class PusherMockWithDisconnect extends PusherMock {
    disconnect() {
        return jest.fn();
    }
}

export default PusherMockWithDisconnect;
