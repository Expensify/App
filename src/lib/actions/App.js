import Ion from '../Ion';
import IONKEYS from '../../IONKEYS';

let currentRedirectTo;
Ion.connect({
    key: IONKEYS.APP_REDIRECT_TO,
    callback: val => currentRedirectTo = val,
});

/**
 * Keep the current route match stored in Ion so other libs can access it
 * Also reset the app_redirect_to in Ion so that if we go back to the current url the state will update
 *
 * @param {object} match
 */
function recordCurrentRoute({match}) {
    Ion.set(IONKEYS.CURRENT_URL, match.url);
    if (match.url === currentRedirectTo) {
        Ion.set(IONKEYS.APP_REDIRECT_TO, '');
    }
}


/**
 * Redirect the app to a new page by updating the state in Ion
 *
 * @param {mixed} url
 */
function redirect(url) {
    const formattedURL = (typeof url === 'string' && url.startsWith('/')) ? url : `/${url}`;
    Ion.set(IONKEYS.APP_REDIRECT_TO, formattedURL);
}

export {
    recordCurrentRoute,
    redirect,
};
