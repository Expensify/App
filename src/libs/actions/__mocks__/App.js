"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.KEYS_TO_PRESERVE = exports.updateLastVisitedPath = exports.createWorkspaceWithPolicyDraftAndNavigateToIt = exports.savePolicyDraftByNewWorkspace = exports.finalReconnectAppAfterActivatingReliableUpdates = exports.beginDeepLinkRedirectAfterTransition = exports.beginDeepLinkRedirect = exports.handleRestrictedEvent = exports.confirmReadyToOpenApp = exports.reconnectApp = exports.openApp = exports.redirectThirdPartyDesktopSignIn = exports.setUpPoliciesAndNavigate = exports.setSidebarLoaded = exports.setLocaleAndNavigate = exports.setLocale = exports.mockValues = exports.getMissingOnyxUpdates = void 0;
var OnyxUpdates = require("@userActions/OnyxUpdates");
var createProxyForObject_1 = require("@src/utils/createProxyForObject");
jest.mock('@libs/actions/OnyxUpdates');
jest.mock('@libs/actions/OnyxUpdateManager/utils/applyUpdates');
var AppImplementation = jest.requireActual('@libs/actions/App');
var setLocale = AppImplementation.setLocale, setLocaleAndNavigate = AppImplementation.setLocaleAndNavigate, setSidebarLoaded = AppImplementation.setSidebarLoaded, setUpPoliciesAndNavigate = AppImplementation.setUpPoliciesAndNavigate, redirectThirdPartyDesktopSignIn = AppImplementation.redirectThirdPartyDesktopSignIn, openApp = AppImplementation.openApp, reconnectApp = AppImplementation.reconnectApp, confirmReadyToOpenApp = AppImplementation.confirmReadyToOpenApp, handleRestrictedEvent = AppImplementation.handleRestrictedEvent, beginDeepLinkRedirect = AppImplementation.beginDeepLinkRedirect, beginDeepLinkRedirectAfterTransition = AppImplementation.beginDeepLinkRedirectAfterTransition, finalReconnectAppAfterActivatingReliableUpdates = AppImplementation.finalReconnectAppAfterActivatingReliableUpdates, savePolicyDraftByNewWorkspace = AppImplementation.savePolicyDraftByNewWorkspace, createWorkspaceWithPolicyDraftAndNavigateToIt = AppImplementation.createWorkspaceWithPolicyDraftAndNavigateToIt, updateLastVisitedPath = AppImplementation.updateLastVisitedPath, KEYS_TO_PRESERVE = AppImplementation.KEYS_TO_PRESERVE;
exports.setLocale = setLocale;
exports.setLocaleAndNavigate = setLocaleAndNavigate;
exports.setSidebarLoaded = setSidebarLoaded;
exports.setUpPoliciesAndNavigate = setUpPoliciesAndNavigate;
exports.redirectThirdPartyDesktopSignIn = redirectThirdPartyDesktopSignIn;
exports.openApp = openApp;
exports.reconnectApp = reconnectApp;
exports.confirmReadyToOpenApp = confirmReadyToOpenApp;
exports.handleRestrictedEvent = handleRestrictedEvent;
exports.beginDeepLinkRedirect = beginDeepLinkRedirect;
exports.beginDeepLinkRedirectAfterTransition = beginDeepLinkRedirectAfterTransition;
exports.finalReconnectAppAfterActivatingReliableUpdates = finalReconnectAppAfterActivatingReliableUpdates;
exports.savePolicyDraftByNewWorkspace = savePolicyDraftByNewWorkspace;
exports.createWorkspaceWithPolicyDraftAndNavigateToIt = createWorkspaceWithPolicyDraftAndNavigateToIt;
exports.updateLastVisitedPath = updateLastVisitedPath;
exports.KEYS_TO_PRESERVE = KEYS_TO_PRESERVE;
var mockValues = {
    missingOnyxUpdatesToBeApplied: undefined,
};
var mockValuesProxy = (0, createProxyForObject_1.default)(mockValues);
exports.mockValues = mockValuesProxy;
var getMissingOnyxUpdates = jest.fn(function (updateIDFrom, updateIDTo) {
    var _a;
    var updates = (_a = mockValuesProxy.missingOnyxUpdatesToBeApplied) !== null && _a !== void 0 ? _a : [];
    if (updates.length === 0) {
        for (var i = updateIDFrom + 1; i <= updateIDTo; i++) {
            updates.push({
                lastUpdateID: i,
                previousUpdateID: i - 1,
            });
        }
    }
    var chain = Promise.resolve();
    updates.forEach(function (update) {
        chain = chain.then(function () {
            if (!OnyxUpdates.doesClientNeedToBeUpdated({ previousUpdateID: Number(update.previousUpdateID) })) {
                return OnyxUpdates.apply(update).then(function () { return undefined; });
            }
            OnyxUpdates.saveUpdateInformation(update);
            return Promise.resolve();
        });
    });
    return chain;
});
exports.getMissingOnyxUpdates = getMissingOnyxUpdates;
