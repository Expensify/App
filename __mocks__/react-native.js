// eslint-disable-next-line no-restricted-imports
import * as ReactNative from 'react-native';
import _ from 'underscore';
import CONST from '../src/CONST';

jest.doMock('react-native', () => {
    let url = 'https://new.expensify.com/';
    const getInitialURL = () => Promise.resolve(url);

    let appState = 'active';
    let count = 0;
    const changeListeners = {};

    // Tests will run with the app in a typical small screen size by default. We do this since the react-native test renderer
    // runs against index.native.js source and so anything that is testing a component reliant on withWindowDimensions()
    // would be most commonly assumed to be on a mobile phone vs. a tablet or desktop style view. This behavior can be
    // overridden by explicitly setting the dimensions inside a test via Dimensions.set()
    let dimensions = CONST.TESTING.SCREEN_SIZE.SMALL;

    return Object.setPrototypeOf(
        {
            NativeModules: {
                ...ReactNative.NativeModules,
                BootSplash: {
                    getVisibilityStatus: jest.fn(),
                    hide: jest.fn(),
                },
                StartupTimer: {stop: jest.fn()},
            },
            Linking: {
                ...ReactNative.Linking,
                getInitialURL,
                setInitialURL(newUrl) {
                    url = newUrl;
                },
            },
            AppState: {
                ...ReactNative.AppState,
                get currentState() {
                    return appState;
                },
                emitCurrentTestState(state) {
                    appState = state;
                    _.each(changeListeners, listener => listener(appState));
                },
                addEventListener(type, listener) {
                    if (type === 'change') {
                        const originalCount = count;
                        changeListeners[originalCount] = listener;
                        ++count;
                        return {
                            remove: () => {
                                delete changeListeners[originalCount];
                            },
                        };
                    }

                    return ReactNative.AppState.addEventListener(type, listener);
                },
            },
            Dimensions: {
                ...ReactNative.Dimensions,
                addEventListener: jest.fn(),
                get: () => dimensions,
                set: (newDimensions) => {
                    dimensions = newDimensions;
                },
            },
        },
        ReactNative,
    );
});
