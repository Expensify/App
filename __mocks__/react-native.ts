// eslint-disable-next-line no-restricted-imports
import * as ReactNative from 'react-native';
import type StartupTimer from '@libs/StartupTimer/types';

const {BootSplash} = ReactNative.NativeModules;

jest.doMock('react-native', () => {
    let url = 'https://new.expensify.com/';
    const getInitialURL = () => Promise.resolve(url);

    let appState: ReactNative.AppStateStatus = 'active';
    let count = 0;
    const changeListeners: Record<number, (state: ReactNative.AppStateStatus) => void> = {};

    // Tests will run with the app in a typical small screen size by default. We do this since the react-native test renderer
    // runs against index.native.js source and so anything that is testing a component reliant on withWindowDimensions()
    // would be most commonly assumed to be on a mobile phone vs. a tablet or desktop style view. This behavior can be
    // overridden by explicitly setting the dimensions inside a test via Dimensions.set()
    let dimensions: Record<string, number> = {
        width: 300,
        height: 700,
        scale: 1,
        fontScale: 1,
    };

    type ReactNativeMock = typeof ReactNative & {
        NativeModules: typeof ReactNative.NativeModules & {
            BootSplash: {
                getVisibilityStatus: typeof BootSplash.getVisibilityStatus;
                hide: typeof BootSplash.hide;
                logoSizeRatio: number;
                navigationBarHeight: number;
            };
            StartupTimer: StartupTimer;
        };
        Linking: typeof ReactNative.Linking & {
            setInitialURL: (newUrl: string) => void;
        };
        AppState: typeof ReactNative.AppState & {
            emitCurrentTestState: (state: ReactNative.AppStateStatus) => void;
        };
    };

    const reactNativeMock: ReactNativeMock = Object.setPrototypeOf(
        {
            NativeModules: {
                ...ReactNative.NativeModules,
                BootSplash: {
                    getVisibilityStatus: jest.fn(),
                    hide: jest.fn(),
                    logoSizeRatio: 1,
                    navigationBarHeight: 0,
                },
                StartupTimer: {stop: jest.fn()},
            },
            Linking: {
                ...ReactNative.Linking,
                getInitialURL,
                setInitialURL(newUrl: string) {
                    url = newUrl;
                },
            },
            AppState: {
                ...ReactNative.AppState,
                get currentState() {
                    return appState;
                },
                emitCurrentTestState(state: ReactNative.AppStateStatus) {
                    appState = state;
                    Object.entries(changeListeners).forEach(([, listener]) => listener(appState));
                },
                addEventListener(type: ReactNative.AppStateEvent, listener: (state: ReactNative.AppStateStatus) => void) {
                    if (type === 'change') {
                        const originalCount = count;
                        changeListeners[originalCount] = listener;
                        ++count;
                        return {
                            remove: () => {
                                delete changeListeners[originalCount];
                            },
                        };
                    }

                    return ReactNative.AppState.addEventListener(type, listener);
                },
            },
            Dimensions: {
                ...ReactNative.Dimensions,
                addEventListener: jest.fn(),
                get: () => dimensions,
                set: (newDimensions: Record<string, number>) => {
                    dimensions = newDimensions;
                },
            },

            // `runAfterInteractions` method would normally be triggered after the native animation is completed,
            // we would have to mock waiting for the animation end and more state changes,
            // so it seems easier to just run the callback immediately in tests.
            InteractionManager: {
                ...ReactNative.InteractionManager,
                runAfterInteractions: (callback: () => void) => callback(),
            },
        },
        ReactNative,
    );

    return reactNativeMock;
});
