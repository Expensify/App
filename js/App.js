/**
 * @format
 * @flow strict-local
 */

import React, {Component} from 'react';
import {SafeAreaView, Text, StatusBar} from 'react-native';
import {Button, TextInput, View} from 'react-native-web';
import * as Store from './store/Store.js';
import {signIn} from './store/actions/SessionActions.js';
import STOREKEYS from './store/STOREKEYS.js';

export default class App extends Component {
  constructor(props) {
    super(props);

    this.submit = this.submit.bind(this);
    this.sessionChanged = this.sessionChanged.bind(this);

    this.state = {
      login: '',
      password: '',
      error: Store.get(STOREKEYS.SESSION, 'error'),
    };
  }

  componentDidMount() {
    // Listen for changes to our session
    Store.subscribe(STOREKEYS.SESSION, this.sessionChanged);
  }

  componentWillUnmount() {
    Store.unsubscribe(STOREKEYS.SESSION, this.sessionChanged);
  }

  /**
   * When the session changes, change which page the user sees
   *
   * @param {object} newSession
   */
  sessionChanged(newSession) {
    this.setState({error: newSession && newSession.error});
  }

  /**
   * When the form is submitted, then we trigger our prop callback
   */
  submit() {
    signIn(this.state.login, this.state.password);
  }

  render() {
    return (
      <>
        <StatusBar barStyle="dark-content" />
        <SafeAreaView>
          <View>
            <Text>Login:</Text>
            <TextInput
              value={this.state.login}
              onChange={(e) => this.setState({login: e.target.value})}
            />
          </View>
          <View>
            <Text>Password:</Text>
            <TextInput
              value={this.state.password}
              onChange={(e) => this.setState({password: e.target.value})}
            />
          </View>
          <View>
            <Button onPress={this.submit}>Log In</Button>
          </View>
        </SafeAreaView>
      </>
    );
  }
}
