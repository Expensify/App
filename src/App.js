import '../wdyr';
import React from 'react';
import {LogBox} from 'react-native';
import {GestureHandlerRootView} from 'react-native-gesture-handler';
import {SafeAreaProvider} from 'react-native-safe-area-context';
import Onyx from 'react-native-onyx';
import CustomStatusBar from './components/CustomStatusBar';
import ErrorBoundary from './components/ErrorBoundary';
import Expensify from './Expensify';
import {LocaleContextProvider} from './components/withLocalize';
import OnyxProvider from './components/OnyxProvider';
import HTMLEngineProvider from './components/HTMLEngineProvider';
import ComposeProviders from './components/ComposeProviders';
import SafeArea from './components/SafeArea';
import * as Environment from './libs/Environment/Environment';
import {WindowDimensionsProvider} from './components/withWindowDimensions';

// For easier debugging and development, when we are in web we expose Onyx to the window, so you can more easily set data into Onyx
if (window && Environment.isDevelopment()) {
    window.Onyx = Onyx;
}

LogBox.ignoreLogs([
    // Basically it means that if the app goes in the background and back to foreground on Android,
    // the timer is lost. Currently Expensify is using a 30 minutes interval to refresh personal details.
    // More details here: https://git.io/JJYeb
    'Setting a timer for a long period of time',
]);

const fill = {flex: 1};

const App = () => (
    <GestureHandlerRootView style={fill}>
        <ComposeProviders
            components={[
                OnyxProvider,
                SafeAreaProvider,
                SafeArea,
                LocaleContextProvider,
                HTMLEngineProvider,
                WindowDimensionsProvider,
            ]}
        >
            <CustomStatusBar />
            <ErrorBoundary errorMessage="NewExpensify crash caught by error boundary">
                <Expensify />
            </ErrorBoundary>
        </ComposeProviders>
    </GestureHandlerRootView>
);

App.displayName = 'App';

export default App;
