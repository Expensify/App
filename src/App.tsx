import {PortalProvider} from '@gorhom/portal';
import React from 'react';
import {LogBox} from 'react-native';
import {GestureHandlerRootView} from 'react-native-gesture-handler';
import {PickerStateProvider} from 'react-native-picker-select';
import {SafeAreaProvider} from 'react-native-safe-area-context';
import '../wdyr';
import ActiveElementRoleProvider from './components/ActiveElementRoleProvider';
import ActiveWorkspaceContextProvider from './components/ActiveWorkspace/ActiveWorkspaceProvider';
import ColorSchemeWrapper from './components/ColorSchemeWrapper';
import ComposeProviders from './components/ComposeProviders';
import CustomStatusBarAndBackground from './components/CustomStatusBarAndBackground';
import CustomStatusBarAndBackgroundContextProvider from './components/CustomStatusBarAndBackground/CustomStatusBarAndBackgroundContextProvider';
import ErrorBoundary from './components/ErrorBoundary';
import HTMLEngineProvider from './components/HTMLEngineProvider';
import InitialURLContextProvider from './components/InitialURLContextProvider';
import {LocaleContextProvider} from './components/LocaleContextProvider';
import OnyxProvider from './components/OnyxProvider';
import PopoverContextProvider from './components/PopoverProvider';
import SafeArea from './components/SafeArea';
import ThemeIllustrationsProvider from './components/ThemeIllustrationsProvider';
import ThemeProvider from './components/ThemeProvider';
import ThemeStylesProvider from './components/ThemeStylesProvider';
import {PlaybackContextProvider} from './components/VideoPlayerContexts/PlaybackContext';
import {VideoPopoverMenuContextProvider} from './components/VideoPlayerContexts/VideoPopoverMenuContext';
import {VolumeContextProvider} from './components/VideoPlayerContexts/VolumeContext';
import {CurrentReportIDContextProvider} from './components/withCurrentReportID';
import {EnvironmentProvider} from './components/withEnvironment';
import {KeyboardStateProvider} from './components/withKeyboardState';
import {WindowDimensionsProvider} from './components/withWindowDimensions';
import Expensify from './Expensify';
import useDefaultDragAndDrop from './hooks/useDefaultDragAndDrop';
import OnyxUpdateManager from './libs/actions/OnyxUpdateManager';
import {ReportAttachmentsProvider} from './pages/home/report/ReportAttachmentsContext';
import type {Route} from './ROUTES';

type AppProps = {
    /** URL passed to our top-level React Native component by HybridApp. Will always be undefined in "pure" NewDot builds. */
    url?: Route;
};

LogBox.ignoreLogs([
    // Basically it means that if the app goes in the background and back to foreground on Android,
    // the timer is lost. Currently Expensify is using a 30 minutes interval to refresh personal details.
    // More details here: https://git.io/JJYeb
    'Setting a timer for a long period of time',
]);

const fill = {flex: 1};

function App({url}: AppProps) {
    useDefaultDragAndDrop();
    OnyxUpdateManager();
    return (
        <InitialURLContextProvider url={url}>
            <GestureHandlerRootView style={fill}>
                <ComposeProviders
                    components={[
                        OnyxProvider,
                        ThemeProvider,
                        ThemeStylesProvider,
                        ThemeIllustrationsProvider,
                        SafeAreaProvider,
                        PortalProvider,
                        SafeArea,
                        LocaleContextProvider,
                        HTMLEngineProvider,
                        WindowDimensionsProvider,
                        KeyboardStateProvider,
                        PopoverContextProvider,
                        CurrentReportIDContextProvider,
                        ReportAttachmentsProvider,
                        PickerStateProvider,
                        EnvironmentProvider,
                        CustomStatusBarAndBackgroundContextProvider,
                        ActiveElementRoleProvider,
                        ActiveWorkspaceContextProvider,
                        PlaybackContextProvider,
                        VolumeContextProvider,
                        VideoPopoverMenuContextProvider,
                    ]}
                >
                    <CustomStatusBarAndBackground />
                    <ErrorBoundary errorMessage="NewExpensify crash caught by error boundary">
                        <ColorSchemeWrapper>
                            <Expensify />
                        </ColorSchemeWrapper>
                    </ErrorBoundary>
                </ComposeProviders>
            </GestureHandlerRootView>
        </InitialURLContextProvider>
    );
}

App.displayName = 'App';

export default App;
