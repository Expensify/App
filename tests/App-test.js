/**
 * @format
 */

import 'react-native';
import React from 'react';

// Note: `react-test-renderer` renderer must be required after react-native.
import renderer from 'react-test-renderer';
import {MemoryRouter} from 'react-router-dom';


import App from '../src/App';
import SignInPage from '../src/pages/SignInPage';

jest.mock('../node_modules/@react-native-community/async-storage',
    () => require('./mocks/@react-native-community/async-storage'));
jest.mock('../node_modules/@react-native-community/netinfo',
    () => require('./mocks/@react-native-community/netinfo'));
jest.mock('../node_modules/@react-native-community/push-notification-ios',
    () => require('./mocks/@react-native-community/push-notification-ios'));
jest.mock('../node_modules/react-native-config',
    () => require('./mocks/react-native-config'));
jest.mock('../node_modules/react-native-image-picker',
    () => require('./mocks/react-native-image-picker'));

it('renders correctly', () => {
    renderer.create(<App />);
});

describe('SampleComponent', () => {
    test('should render', () => {
        const component = renderer
            .create(
                <MemoryRouter>
                    <SignInPage />
                </MemoryRouter>
            )
            .toJSON();

        expect(component).toMatchSnapshot();
    });
});
