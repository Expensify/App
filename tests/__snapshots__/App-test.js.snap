// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SampleComponent should render 1`] = `null`;
